### DISCUSSION TITLE
Reading a binary RSA key from memory?

### DISCUSSION THREAD

**Discussion by elseifthen** [lsh123/xmlsec#460] [Q&A] [answered]
I wanted to store an RSA key in binary format in the code and retrieve it via e.g. `xmlSecKeyReadMemory(xmlSecKeyDataRsaId, c_key, i_len)`  but that doesn't seem to work as the binRead function is not associated with the type `xmlSecKeyDataRsaId` . I managed to load the key in pem format via `xmlSecCryptoAppKeyLoadMemory(c_key, i_len, xmlSecKeyDataFormatPem, NULL, NULL, NULL)` but my question is: is it possible at all to load a _binary_ version of the RSA key from memory or is there a fundamental problem with that?
N.B.: I succeeded in loading a _symmetric_ (AES) key from memory.

**Comment by lsh123** [comment, marked as answer]
First, this functionality is crypto library specific. If you are using default OpenSSL, then multiple RSA key formats are supported -- PEM, DER, ...: https://github.com/lsh123/xmlsec/blob/fd87b54f5e43f55b997cc394ca0a59f4ae7df54a/src/openssl/app.c#L304 

Note that RSA key has multiple components while AES/DES keys have only one. Thus, you need to know how these components are stored in memory to correctly read the key.

You can also use something like xmlSecOpenSSLEvpKeyDataAdoptEvp() function to use an RSA key stored in OpenSSL EVP key which can be initialized / loaded using any relevant OpenSSL functions (e.g. EVP key can be on a hardware device).

Aleksey

**Comment by elseifthen** [comment]
Thanks for the hints! Using the `xmlSecKeyDataFormatDer` format, loading the binary version of the RSA key works with
`xmlSecCryptoAppKeyLoadMemory(c_key, i_len, xmlSecKeyDataFormatDer, NULL, NULL, NULL)`
